lib/queues.js

```diff
@@ -320,16 +319,14 @@ class Queue {
       const backoff = ifProduction(RESERVE_BACKOFF);
       const wait    = (process.env.NODE_ENV === 'test' ? 0 : msToSec(RESERVE_WAIT));
 
-      this._server.configPromise
-        .then(config => {
-          const maxJobs = config.concurrency - this._activeJobs;
-
+      this._server.getReserveCapacity()
+        .then(maxJobs => {
           if (maxJobs > 0) {
-            debug('Waiting for jobs on queue %s (max %s)', this.name, maxJobs);
+            debug('Waiting for jobs on queue %s (at most %s)', this.name, maxJobs);
             return this._reserveJob(client, wait, maxJobs);
           } else {
-            debug('%s out of %s %s jobs running, nothing to do', this._activeJobs, config.concurrency, this.name);
-            return Bluebird.delay(500).then(() => []);
+            // No capacity to consume jobs right now.
+            return Bluebird.delay(10).then(() => []);
```

Comment:
The debug message is useful, don't want to lose it.

The problem is, this code is spinning waiting to have some processing capacity. There is no right value for delay: too long means idle time, too short is eating up CPU and logs. Where in fact, all this code needs to do is wait to be notified when active jobs is less than allowed concurrency.